conda/lock.py

```diff
@@ -20,52 +20,79 @@
+If you are sure that conda is not running, remove it and try again.
+You can also use: $ conda clean --lock
+"""
 
 stdoutlog = logging.getLogger('stdoutlog')
+log = logging.getLogger(__name__)
+
+
+def touch(file_name, times=None):
+    """ Touch function like touch in Unix shell
+    :param file_name: the name of file
+    :param times: the access and modified time
+    Examples:
+        touch("hello_world.py")
+    """
+    with open(file_name, 'a'):
+        os.utime(file_name, times)
 
 
-class Locked(object):
+class FileLock(object):
     """
     Context manager to handle locks.
     """
-    def __init__(self, path, retries=10):
-        self.path = path
-        self.end = "-" + str(os.getpid())
-        self.lock_path = os.path.join(self.path, LOCKFN + self.end)
+    def __init__(self, file_path, retries=10):
+        """
+        :param filepath: The file or directory to be locked
+        :param retries: max number of retries
+        :return:
+        """
+        self.file_path = abspath(file_path)
         self.retries = retries
 
     def __enter__(self):
-        # Keep the string "LOCKERROR" in this string so that external
-        # programs can look for it.
-        lockstr = ("""\
-    LOCKERROR: It looks like conda is already doing something.
-    The lock %s was found. Wait for it to finish before continuing.
-    If you are sure that conda is not running, remove it and try again.
-    You can also use: $ conda clean --lock\n""")
-        sleeptime = 1
 
-        for _ in range(self.retries):
-            if os.path.isdir(self.lock_path):
-                stdoutlog.info(lockstr % self.lock_path)
-                stdoutlog.info("Sleeping for %s seconds\n" % sleeptime)
+        sleep_time = 1
+        self.lock_path = "{0}.pid{1}.{2}".format(self.file_path, os.getpid(), LOCKFN)
+        lock_glob_str = "{0}.pid*.{1}".format(self.file_path, LOCKFN)
+        last_glob_match = None
 
-                time.sleep(sleeptime)
-                sleeptime *= 2
+        for _ in range(self.retries):
+            # search, whether there is process already locked on this file
+            glob_result = glob(lock_glob_str)
+            if glob_result:
+                stdoutlog.info(LOCKSTR % glob_result[0])
+                stdoutlog.info("Sleeping for %s seconds\n" % sleep_time)
+
+                time.sleep(sleep_time/10)
+                sleep_time *= 2
+                last_glob_match = glob_result
             else:
-                os.makedirs(self.lock_path)
+                # Just for unittest, should never happen in real-world application
+                if not isdir(dirname(self.lock_path)):
+                    os.makedirs(dirname(self.file_path))
+                # create a lock
+                touch(self.lock_path)
                 return self
 
         stdoutlog.error("Exceeded max retries, giving up")
-        raise LockError(lockstr % self.lock_path)
+        raise LockError(LOCKSTR % last_glob_match[0])
```

Comment:
We should also now modify `LOCKSTR` so we can pass in the full `last_glob_match`.  It's probably just changing the english to make it make sense.